pkg/lambci-lambda/service.go

```diff
@@ -0,0 +1,278 @@
+		return nil, err
+	}
+	if err := createDirIfNotExists(layersDir); err != nil {
+		return nil, err
+	}
+
+	env := filepath.Join(service.homeDir, envFile)
+	if _, err := os.Stat(env); err != nil {
+		_, err = service.Exec(fmt.Sprintf("touch %s", env))
+		if err != nil {
+			return nil, errors.Wrap(err, "error creating lambci env file")
+		}
+	}
+
+	for name, config := range config.GetStringMap("runtimes") {
+
+		runtimeConfig := config.(map[string]interface{})
+		layerConfig := runtimeConfig["layers"].([]interface{})
+
+		layers := make([]string, len(layerConfig))
+		for i := 0; i < len(layerConfig); i++ {
+			layers[i] = layerConfig[i].(string)
+		}
+
+		service.runtimes[name] = layers
+	}
+
+	return service, nil
+}
+
+// Package up everything needed to install the function but don't actually
+// install it to the service. rawDir may be assumed to be a unique path for
+// this function. The package location should be determinsitically derived
+// from the rawDir path.
+// Returns: Path to the newly created package
+func (service *lambciLambda) Package(rawDir string) (string, error) {
+
+	// for lambci-lambda there is no need to create zip files
+	return rawDir, nil
+}
+
+// Install a function to the desired FaaS service. It is assumed that
+// Package() has already been called on this rawDir. The name of rawDir is
+// also the name of the function.
+func (service *lambciLambda) Install(rawDir string, env map[string]string, runtime string) error {
+
+	if runtime == "" {
+		runtime = service.defaultRuntime
+	}
+
+	// we allow no runtime for lambci as it could be provided by the container
+	if runtime != "" {
+		if _, exists := service.runtimes[runtime]; !exists {
+			return errors.Errorf("runtime '%s' does not exist in configuration", runtime)
+		}
+	}
+
+	// remove old layer
+	_, err := service.Exec(fmt.Sprintf("find %s -mindepth 1 -maxdepth 1 -exec rm -r {} +", filepath.Join(service.homeDir, runtimeDir)))
+	if err != nil {
+		return errors.Wrap(err, "error removing old layer")
+	}
+
+	// install new layer
+	if runtime != "" {
+		for _, layer := range service.runtimes[runtime] {
+			_, err := service.Exec(fmt.Sprintf("cp -r %s %s", filepath.Join(service.homeDir, layersDir, layer, "*"), filepath.Join(service.homeDir, runtimeDir)))
+			if err != nil {
+				return errors.Wrapf(err, "error installing layer '%s'", layer)
+			}
+		}
+	}
+
+	// remove old task
+	_, err = service.Exec(fmt.Sprintf("find %s -mindepth 1 -maxdepth 1 -exec rm -r {} +", filepath.Join(service.homeDir, taskDir)))
+	if err != nil {
+		return errors.Wrap(err, "error removing old task")
+	}
+
+	// install new task
+	_, err = service.Copy(filepath.Join(rawDir, "*"), filepath.Join(service.homeDir, taskDir))
+	if err != nil {
+		return errors.Wrap(err, "error installing function")
+	}
+
+	// retrieve process id of running lambda docker image
+	pid, err := service.Exec("ps ax | grep \"LAMBDA\" | grep -v entr | grep -v grep | awk \"{print $1}\"")
```

Comment:
I seem to have missed where the docker container is launched. Is the user supposed to do it before calling srk?

Reply:
Yes. We should consider to integrate the launch of the service provider in the future.